src/backend/distributed/connection/locally_reserved_shared_connection_stats.c

```diff
@@ -0,0 +1,508 @@
+/*-------------------------------------------------------------------------
+ *
+ * locally_reserved_shared_connection_stats.c
```

Comment:
The naming is getting a little trippy.

We should maybe consider renaming the shard connection counters, e.g. "connection budget" and "connection budget reservation". Probably a separate PR.

Reply:
I agree  that `locally_reserved_shared_connection_stats` is not a good name :D 
 

On the  other hand, `shared connection counters` doesn't feel like a budget as it holds the connections that are actually established.

I'll try to come up with  some alternatives

Reply:
Actually,  `shared  connections` can  be  qualified as budget as you noted.

>  "connection budget" and "connection budget reservation"

So, the above sounds good to me as well.

> Probably a separate PR.

Do you mean after we merge this, not before, right?

Reply:
> Do you mean after we merge this, not before, right?

right